torchtitan/datasets/experimental_datasets.py

```diff
@@ -0,0 +1,1212 @@
+            cur += docrange
+            if cur > i:
+                return shardid, docrange, min_d
+
+    def _get_reader(self, path, newpath, reader):
+        """
+        If new filepath does not match the current one,
+        open a new reader on that filepath (pull file on demand)
+        """
+        if newpath != path:
+            del reader
+            if self.verbose:
+                logging.info(f"Worker {self.rank} opening new file {newpath}")
+            reader = pa.ipc.open_file(newpath)
+            path = newpath
+        return path, reader
+
+    def _construct_chunk(self, j, doc, n_chunks):
+        """
+        Grab a chunk of the desired size from the pyarrow document,
+        avoiding unnecessary overhead in case of large docs
+        """
+        start_index = j * self.chunksize
+        n_pull = self.chunksize
+        if self.bos is not None:
+            if j == 0:
+                n_pull -= 1
+            else:
+                start_index -= 1
+        chunk = doc.slice(start_index, n_pull).to_pylist()
+        self.tokens_seen += len(chunk)
+        # Add bos/eos tokens if needed
+        if self.bos is not None and j == 0:
+            chunk = [self.bos] + chunk
+        if j == n_chunks - 1:
+            chunk = chunk + [self.eos]
+        return chunk
+
+    def _random_map_docid(self, size):
+        """
+        Given size of document pool, use saved state (prior index) to generate the next index via LCG.
+        Implements within-shard document shuffling without materializing any large doc lists.
+        """
+        m = 2 ** math.ceil(math.log2(size))  # Round up to nearest power of 2
+        a = 5  # A,C values known to work well with powers of 2 (Knuth, 1997, 3.2.1.3)
+        c = (self.rank + self.seed) * 2 + 1
+        state = self.lcg_state
+        while True:
+            state = (a * state + c) % m
+            if state < size:
+                return state
+
+    def __iter__(self):
+        docset_offset = self.docset_index
+        lcg_offset = self.lcg_state
+        residual_chunks = self.chunk_index + 1  # pick up AFTER where the ckp left off
+        ndocs = self._len
+        path = ""
+        reader = None
+        while True:
+            # Iterate through docs, starting at desired offset
+            for i in range(ndocs):
+                doc_index = (docset_offset + i) % ndocs
+
+                # Update stats
+                if doc_index == 0:
+                    self.epochs_seen += 1
+                self.docset_index = doc_index
+                # Map doc id to shard, id in file
+                shardid, docrange, mindoc = self._get_docid(doc_index)
+
+                # Read doc
+                newpath = os.path.join(self.data, shardid)
+                path, reader = self._get_reader(path, newpath, reader)
+                # Map id in range of owned docs to new (consistently) shuffled id
+                doclcg = self._random_map_docid(docrange)
```

Comment:
Just to confirm, once a shard is chosen, all the docs within that shard is visited in random order but that shard is exhausted before moving to the next shard?

Reply:
Technically, once a shard is chosen, all the docs within the _fraction of the shard owned by this worker_ are visited. But yes that's broadly correct.